tsinfer/inference.py

```diff
@@ -452,6 +460,136 @@ def match_samples(
+    ancestor_data file) have been explicitly set by the user, e.g. by using the
+    ``time`` parameter of ``SampleData.add_sites()``, or by setting ``use_times=True`` if
+    creating a sample data file via ``tsinfer.SampleData.from_tree_sequence()``.
+
+    If ``create_sample_nodes`` is True, the inserted ancestors will be flagged with
+    tsinfer.NODE_IS_TRUE_SAMPLE_ANCESTOR, and the individuals associated with the
+    non-contemporaneous samples will be present in the resulting ancestors tree sequence.
+    This will create internal sample nodes in the final tree sequence produced by
+    match_samples. If create_sample_nodes is False, the inserted ancestors are treated as
+    *proxies* for the true samples (i.e. close, unrecombined relatives of the
+    true samples) and the nodes will not be contained within an individual in the final
+    tree sequence.
+
+    Returns an ancestors tree sequence suitable for input into match_samples. If
+    ``create_sample_nodes`` is True, you probably want to exclude any noncontemporaneous
+    samples during the match_samples phase, otherwise they risk being duplicated in the
+    final tree sequence.
+    """
+
+    # First find the sites from the samples file used in the built ancestors
+    sites_used = np.isin(sample_data.sites_position[:], ancestor_data.sites_position[:])
+    sites_used_time = sample_data.sites_time[:][sites_used]
+    if np.any(sites_used_time == constants.TIME_UNSPECIFIED):
+        raise ValueError(
+            "Noncontemporaneous samples can only be ancestors if site times are present"
+        )
+    samples_population = sample_data.samples_population[:]
+
+    # Find the non-contemporaneous samples and group them by individuals_time
+    samples_individual = sample_data.samples_individual[:]
+    noncontemp_samples = np.where(
+        np.isin(samples_individual, np.where(sample_data.individuals_time[:] > 0)[0])
+    )[0]
+    samples_by_timepoint = collections.defaultdict(list)
+    for s in noncontemp_samples:
+        timepoint = sample_data.individuals_time[samples_individual[s]]
+        samples_by_timepoint[timepoint].append(s)
+
+    logger.info("Creating new ancestors file with added ancestors")
+    ancestor_to_sample_map = {}
+    inferred_ancestors = ancestor_data.ancestors()
+    current_anc = next(inferred_ancestors)
+    with formats.AncestorData(sample_data=sample_data) as new_data:
+        for insert_time, sample_ids in sorted(samples_by_timepoint.items()):
+            while current_anc is not None and current_anc.time > insert_time:
+                new_data.add_ancestor(**attr.asdict(current_anc, filter=exclude_id))
+                current_anc = next(inferred_ancestors, None)
+            for s in sample_ids:
+                logger.debug(f"Adding sample {s} as an ancestor")
+                haplotype = sample_data.sites_genotypes[:, s][sites_used]
+                if np.any(haplotype == tskit.MISSING_DATA):
```

Comment:
Almost certain that this can be removed. @jeromekelleher , missing data in ancestors is supported, is that right? I can't find a test case in test_inference.py though

Reply:
Should work fine, I probably just didn't test it explicitly.

Reply:
I'm working through it in #306